projectiles/projectile_effect.gd

```diff
@@ -0,0 +1,75 @@
+class_name ProjectileEffect
+extends Node
+# A ProjectileEffect is a set of post-processors meant to be run on every instance of a Projectile.
+# This by itself does nothing but is meant to be extended by various types of projectile effect.
+# Each effect will compound on the others. Each function will be called by the Projectile scene,
+#    in order of Effect Priority
+# TODO:: Effect Priority as an enum? (to discourage "priority: 9999999")
+
+# Each modifier is a combination of a double value and a modification type.
+# This is to indicate if a modification is additive or multiplicitive.
+# The default for all modifiers is MULTIPLICITIVE : 1, ultimately resulting in no effect.
+# (To subtract and divide, use negative values and decimals < 1)
+enum MODIFICATION_TYPE {
+	ADD,
+	MULTIPLY
+}
+@export var effect_priority = -1
+@export var speed_modifier_type = MODIFICATION_TYPE.MULTIPLY
+@export var speed_modifier = 1
+@export var damage_modifier_type = MODIFICATION_TYPE.MULTIPLY
+@export var damage_modifier = 1
+@export var scale_modifier_type = MODIFICATION_TYPE.MULTIPLY
+@export var scale_modifier = 1
+
+
+# Modify the logic that adds the projectile as a child to the scene.
+# This can be used to change spawn locations, quantities, etc.
+func modify_creation(owner: Node, projectile_effects: Array, transform: Transform2D):
```

Comment:
:wine_glass: :cheese: 

I'm pretty sure the resulting type here is "Packed Scene", which currently [cannot be typed better](https://github.com/godotengine/godot-proposals/issues/8565)

I flip-flopped between the actual node type (ProjectileEffect) and PackedScenes and some other ideas so many times that changing every type annotation was incredibly cumbersome, so I've opted to leave this as just `Array` until we flesh things out and settle on something that works and won't be changed.

There's also no support for Type Variables -- I figured having a Globals script that contains this type so we can just say `var EffectType = Array[PackedScene]` and use that so that making a change would be trivial, but can't find a way to do it with the current gdscript version.

Reply:
💬 Some interesting thoughts in this forum post on accessing more info about types when using PackedScene. A custom resource type is an interesting approach. Nothing for now though, just food for collective thought.

Reply:
Classic me. No link, and now I can't find it again 🤦